lib/srv/db/common/auth.go

```diff
@@ -347,22 +360,58 @@ func (a *dbAuth) GetTLSConfig(ctx context.Context, sessionCtx *Session) (*tls.Co
 	if sessionCtx.Database.IsRDS() || sessionCtx.Database.IsRedshift() || sessionCtx.Database.IsCloudSQL() || sessionCtx.Database.IsAzure() {
 		return tlsConfig, nil
 	}
-	// Otherwise, when connecting to an onprem database, generate a client
-	// certificate. The database instance should be configured with
-	// Teleport's CA obtained with 'tctl auth sign --type=db'.
-	cert, cas, err := a.getClientCert(ctx, sessionCtx)
-	if err != nil {
-		return nil, trace.Wrap(err)
-	}
-	tlsConfig.Certificates = []tls.Certificate{*cert}
-	for _, ca := range cas {
-		if !tlsConfig.RootCAs.AppendCertsFromPEM(ca) {
-			return nil, trace.BadParameter("failed to append CA certificate to the pool")
+
+	// If the root CA has been already added. Skip the client certificate generation.
+	if !rootsAdded {
+		// Otherwise, when connecting to an onprem database, generate a client
+		// certificate. The database instance should be configured with
+		// Teleport's CA obtained with 'tctl auth sign --type=db'.
+		cert, cas, err := a.getClientCert(ctx, sessionCtx)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		tlsConfig.Certificates = []tls.Certificate{*cert}
+		for _, ca := range cas {
+			if !tlsConfig.RootCAs.AppendCertsFromPEM(ca) {
+				return nil, trace.BadParameter("failed to append CA certificate to the pool")
+			}
 		}
 	}
+
+	if dbTLSConfig.Mode == types.DatabaseTLSMode_INSECURE {
+		// Accept any certificate provided by database.
+		tlsConfig.InsecureSkipVerify = true
+	} else if dbTLSConfig.Mode == types.DatabaseTLSMode_VERIFY_CA {
+		// Base on https://github.com/golang/go/blob/master/src/crypto/tls/example_test.go#L193-L208
+		// Set InsecureSkipVerify to skip the default validation we are
+		// replacing. This will not disable VerifyConnection.
+		tlsConfig.InsecureSkipVerify = true
+		tlsConfig.VerifyConnection = VerifyConnectionFunc(tlsConfig.RootCAs, dbTLSConfig.ServerName)
+	}
+
 	return tlsConfig, nil
 }
 
+// VerifyConnectionFunc returns a certificate validation function. serverName if empty will skip the hostname validation.
+func VerifyConnectionFunc(RootCAs *x509.CertPool, serverName string) func(cs tls.ConnectionState) error {
```

Comment:
```suggestion
func VerifyConnectionFunc(rootCAs *x509.CertPool, serverName string) func(cs tls.ConnectionState) error {
```

Also, does this method need to be exported?